### PR TITLE
DrivAerML: layer-wise learning rate decay (LLRD) across transformer layers

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DM Transolver uses a single learning rate (5e-4) for all 4 transformer layers. However, different layers learn different abstractions — early layers capture low-level spatial patterns, later layers build high-level flow structures. **Layer-wise Learning Rate Decay (LLRD)** assigns progressively lower learning rates to earlier layers, allowing later layers to adapt faster while earlier layers change more conservatively.

LLRD is standard practice in NLP fine-tuning (Clark et al. ELECTRA, He et al. DeBERTa) where it consistently improves over uniform LR. The mechanism: early layers learn stable, reusable features that shouldn't change rapidly, while later layers are more task-specific and need faster adaptation. This is directly analogous to the DM Transolver where early layers build geometric representations and later layers predict surface quantities.

With only 4 layers, LLRD creates a modest but meaningful gradient: layer 0 (closest to input) gets lr × decay^3, layer 3 (closest to output) gets the full lr. With decay=0.8 and lr=5e-4: layer 0=2.56e-4, layer 1=3.2e-4, layer 2=4e-4, layer 3=5e-4. This is a gentle modulation, not a dramatic change.

### Implementation

1. **Add CLI flag:**
   ```python
   parser.add_argument('--layer-lr-decay', type=float, default=1.0,
                       help='Per-layer LR decay factor (1.0=uniform, 0.8=recommended). Layer i gets lr * decay^(num_layers-1-i)')
   ```

2. **Create parameter groups with per-layer LR** (replace the single optimizer construction):
   ```python
   if args.layer_lr_decay < 1.0:
       param_groups = []
       num_layers = args.model_layers
       for i, layer in enumerate(model.transformer_layers):
           layer_lr = args.lr * (args.layer_lr_decay ** (num_layers - 1 - i))
           param_groups.append({'params': layer.parameters(), 'lr': layer_lr})
       # Non-transformer params (input projection, output head, Fourier, etc) get full LR
       transformer_params = set()
       for layer in model.transformer_layers:
           transformer_params.update(layer.parameters())
       other_params = [p for p in model.parameters() if p not in transformer_params]
       param_groups.append({'params': other_params, 'lr': args.lr})
       
       optimizer = torch.optim.AdamW(param_groups, weight_decay=args.weight_decay)
   else:
       optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)
   ```

3. **Ensure the cosine scheduler still works.** The scheduler should apply to the base LR of the optimizer. With parameter groups, `CosineAnnealingLR` will scale each group's LR proportionally — verify this by printing LRs at a few steps during smoke test.

### Trials

**Smoke test first (3 epochs)** — verify loss finite, print per-layer LRs at step 0 and step 30 (first restart) to confirm they scale correctly.

**Trial 1 — decay=0.8 (moderate gradient):**
Layer LRs: [2.56e-4, 3.2e-4, 4e-4, 5e-4]
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --layer-lr-decay 0.8 \
  --wandb_group dm-layer-lr-decay
```

**Trial 2 — decay=0.65 (stronger gradient):**
Layer LRs: [1.37e-4, 2.11e-4, 3.25e-4, 5e-4]
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --layer-lr-decay 0.65 \
  --wandb_group dm-layer-lr-decay
```

Report `val_primary/surface_rel_l2_pct` for both trials. Target: beat **3.833%**.

## Baseline

Current best DrivAerML metrics (PR #3072):

| Metric | Value | Epoch |
|--------|-------|-------|
| val_primary/surface_rel_l2_pct | **3.833%** | 511 |
| test_primary/surface_rel_l2_pct | **4.685%** | 511 |

- W&B run: `ncl1dh88`
- Config: 4L/512d/8H, AdamW lr=5e-4, T_max=30, gc=0.5, EMA=0.9995, no WD, Fourier

Reproduce:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nobara:

## Interim Update (epoch ~110 / ~480 expected)

Both LLRD trials are running on GPUs 0 and 1. Implementation verified via 3-epoch smoke test.

### Per-layer LR configuration
| Layer | Trial 1 (decay=0.8) | Trial 2 (decay=0.65) |
|-------|--------------------|--------------------|
| Layer 0 (early) | 2.56e-4 | 1.37e-4 |
| Layer 1 | 3.20e-4 | 2.11e-4 |
| Layer 2 | 4.00e-4 | 3.25e-4 |
| Layer 3 (late) | 5.00e-4 | 5.00e-4 |
| Other params | 5.00e-4 | 5.00e-4 |

### Progress so far (epoch 110)
| Metric | Trial 1 (decay=0.8) | Trial 2 (decay=0.65) | Baseline |
|--------|--------------------|--------------------|----------|
| best val_primary/surface_rel_l2_pct | **6.14%** | **6.56%** | **3.833%** |

Both runs converging smoothly. Trial 1 (decay=0.8) has a slight edge. Runs will continue until the 360-minute timeout (~480 epochs). Will post final results when complete.

W&B runs: `6ekne3bv` (decay=0.8), `0bf7g3ck` (decay=0.65)

**Comment by tcapelle** [comment]
STUDENT nobara:

## Results

### Summary

| Metric | Trial 1 (decay=0.8) | Trial 2 (decay=0.65) | Baseline |
|--------|--------------------|--------------------|----------|
| **best val_primary/surface_rel_l2_pct** | **6.003%** (epoch 117) | **4.446%** (epoch 292) | **3.833%** (epoch 511) |
| Status | Diverged at epoch 120 | Diverged at epoch ~298 | Stable |
| Peak memory | ~6 GB | ~6 GB | — |

**Neither trial beats the baseline.**

### Per-layer LR configuration (4 transformer layers)

| Layer | Trial 1 (decay=0.8) | Trial 2 (decay=0.65) |
|-------|--------------------|--------------------|
| Layer 0 (early/deepest) | 2.56e-4 (0.512× base) | 1.37e-4 (0.274× base) |
| Layer 1 | 3.20e-4 (0.640× base) | 2.11e-4 (0.423× base) |
| Layer 2 | 4.00e-4 (0.800× base) | 3.25e-4 (0.650× base) |
| Layer 3 (late/top) | 5.00e-4 (1.000× base) | 5.00e-4 (1.000× base) |
| Other params | 5.00e-4 (1.000× base) | 5.00e-4 (1.000× base) |

### Convergence trajectory (Trial 2, best performing)

| Epoch | best_val | Cosine cycle |
|-------|---------|-------------|
| 30 | 15.401% | End of cycle 1 |
| 60 | 9.952% | End of cycle 2 |
| 90 | 7.294% | End of cycle 3 |
| 120 | 6.239% | End of cycle 4 |
| 150 | 5.530% | End of cycle 5 |
| 180 | 5.008% | End of cycle 6 |
| 210 | 4.815% | End of cycle 7 |
| 240 | 4.583% | End of cycle 8 |
| 270 | 4.536% | End of cycle 9 |
| **292** | **4.446%** | **Best (cycle 10, pre-restart)** |
| 300 | 4.446% (frozen) | **Diverged at cosine restart** |

### Commands

```bash
# Trial 1
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 --layer-lr-decay 0.8 --agent nobara --wandb-name "nobara/llrd-decay08" --wandb-group "dm-layer-lr-decay"

# Trial 2
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --ema-decay 0.9995 --layer-lr-decay 0.65 --agent nobara --wandb-name "nobara/llrd-decay065" --wandb-group "dm-layer-lr-decay"
```

### W&B runs
- Trial 1 (decay=0.8): `6ekne3bv`
- Trial 2 (decay=0.65): `0bf7g3ck`

### What happened

**LLRD does not beat the baseline**, and reveals a critical stability issue with cosine restarts.

**Divergence pattern**: Both trials catastrophically diverged at cosine restart boundaries (where LR jumps back to max). Trial 1 (decay=0.8) diverged at epoch 120 (4th restart), Trial 2 (decay=0.65) diverged at epoch 298 (10th restart). In both cases, grad norms spiked from ~0.2 to infinity, loss went from ~0.005 to ~0.15+, and validation error jumped 4× within 2-3 epochs. The EMA model was unable to recover.

**Why stronger decay helps stability**: With decay=0.65, deeper layers get much lower LRs (0.274× base), making them less susceptible to the LR shock at restarts. With decay=0.8, deeper layers are at 0.512× base — still too high for restart stability. This explains why Trial 2 survived 10 restarts vs Trial 1's 4.

**Why it ultimately fails**: Even with stronger decay, the top transformer layer and all "other" params still get the full base LR. At restart, these parameters experience the full LR jump, eventually accumulating enough instability to cascade through all layers.

**Before divergence, Trial 2 was trending toward the baseline** — at 4.446% (epoch 292) and still improving ~0.05% per cycle. Without the restart instability, it may have reached ~4.0-4.2% by epoch 480. The hypothesis itself (differential learning rates across layers) shows promise, but is incompatible with the current cosine restart schedule.

### Suggested follow-ups

1. **LLRD + cosine decay without restarts** (`cosine_t_max=999`): The most promising next step. LLRD showed clear value (4.446% in 292 epochs) but was killed by restarts. A monotonically decreasing LR schedule would avoid the restart shock entirely.

2. **LLRD + warmup restarts**: Modify the scheduler to add a brief warmup period (e.g., 5 epochs of linear ramp-up) after each cosine restart, softening the LR shock.

3. **LLRD + reduced max LR**: Lower the peak LR (e.g., 3e-4 instead of 5e-4) to reduce the magnitude of the restart shock while keeping the per-layer differentiation.

4. **LLRD + gradient norm warmup at restarts**: Clip gradients more aggressively for the first few epochs after a restart to prevent the cascading divergence.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3271 because ADVISOR: Closing — LLRD diverges at cosine restart boundaries. Trial 2 (decay=0.65) best: 4.446% at ep292, then catastrophic divergence at restart 10. Trial 1 (decay=0.8) diverged even earlier at ep120. The per-layer LR differentiation is incompatible with T_max=30 cosine restarts — the full-LR top layers destabilize at each restart. The convergence rate before divergence (~0.05%/cycle) wouldn't beat baseline even without instability. DM's flat architecture (4L) doesn't benefit enough from LLRD to justify the stability cost..
